src/gradualizer.erl

```diff
@@ -227,11 +276,82 @@ add_source_file_and_forms_to_opts(File, Forms, Opts) ->
         false -> Opts1
     end.
 
-%% Extract -gradualizer(Options) from AST
+%% Extract `-gradualizer(Options)' from AST
 -spec options_from_forms(gradualizer_file_utils:abstract_forms()) -> options().
 options_from_forms([{attribute, _L, gradualizer, Opts} | Fs]) when is_list(Opts) ->
     Opts ++ options_from_forms(Fs);
 options_from_forms([{attribute, _L, gradualizer, Opt} | Fs]) ->
     [Opt | options_from_forms(Fs)];
 options_from_forms([_F | Fs]) -> options_from_forms(Fs);
 options_from_forms([]) -> [].
+
+%% @doc Return a Gradualizer type for the passed in Erlang type definition.
+%%
+%% ```
+%% > gradualizer:type("[integer()]").
+%% {type,0,list,[{type,0,integer,[]}]}
+%% '''
+-spec type(string()) -> typechecker:type().
+type(Type) ->
+    typelib:remove_pos(typelib:parse_type(Type)).
+
+%% @see env/2
+-spec env() -> typechecker:env().
+env() ->
+    env([]).
+
+%% @see env/2
+-spec env(gradualizer:options()) -> typechecker:env().
+env(Opts) ->
+    env("", Opts).
+
+%% @doc Create a type checker environment populated by types defined in a source code snippet
+%% via `-type ...' and `-record(...)' attributes.
+%%
+%% ```
+%% > rr(typechecker).
+%% > gradualizer:env("-type a() :: integer().", []).
+%% #env{fenv = #{},imported = #{},venv = #{},
+%%      tenv = #{module => undefined,records => #{},
+%%               types => #{{a,0} => {[],{type,0,integer,[]}}}},
+%%      infer = false,verbose = false,exhaust = true,
+%%      clauses_stack = [],union_size_limit = 30,
+%%      current_spec = none}
+%% > gradualizer:env("-record(r, {f}).", []).
+%% #env{
+%%     fenv = #{},imported = #{},venv = #{},
+%%     tenv =
+%%         #{module => undefined,
+%%           records =>
+%%               #{r =>
+%%                     [{typed_record_field,
+%%                          {record_field,0,{atom,0,f},{atom,1,undefined}},
+%%                          {type,0,any,[]}}]},
+%%           types => #{}},
+%%     infer = false,verbose = false,exhaust = true,
+%%     clauses_stack = [],union_size_limit = 30,
+%%     current_spec = none}
+%% '''
+-spec env(string(), gradualizer:options()) -> typechecker:env().
+env(ErlSource, Opts) ->
+    Forms = gradualizer_lib:ensure_form_list(merl:quote(lists:flatten(ErlSource))),
+    ErlParseForms = lists:map(fun erl_syntax:revert/1, Forms),
+    ParseData = typechecker:collect_specs_types_opaques_and_functions(ErlParseForms),
+    typechecker:create_env(ParseData, Opts).
+
+%% @see type_of/2
+-spec type_of(string()) -> typechecker:type().
+type_of(Expr) ->
+    type_of(Expr, env([infer])).
+
+%% @doc Infer type of an Erlang expression.
+%%
+%% ```
+%% > gradualizer:type_of("[ a || _ <- lists:seq(1, 5) ]").
+%% {type,0,list,[{atom,0,a}]}
+%% '''
+-spec type_of(string(), typechecker:env()) -> typechecker:type().
+type_of(Expr, Env) ->
```

Comment:
Add an example where type_of/2 and env/1 are used together? They are only useful together in practice, right? `gradualizer:type_of("...", gradualizer:env("..."))`.

Btw, don't we need to expose pretty-printing of types too for this to be useful without reading abstract internal representations?

Reply:
> They are only useful together in practice, right?

Indeed. Example added.

> don't we need to expose pretty-printing of types too?

Pretty-printing is already exported, see the module doc at the top:

```
> gradualizer:type_of("fun (A) -> #{tag => my_map, list_of_as => [ A || _ <- lists:seq(1, 5) ]} end").
{type,0,'fun',
      [{type,0,product,[{type,0,any,[]}]},
       {type,0,map,
             [{type,0,map_field_assoc,[{atom,0,tag},{atom,0,my_map}]},
              {type,0,map_field_assoc,
                    [{atom,0,list_of_as},{type,0,list,[{type,0,any,[]}]}]}]}]}
> typelib:pp_type(v(-1)).
"fun((any()) -> #{tag => my_map, list_of_as => [any()]})"
```

Reply:
I've also realised that defining variable bindings and variable typings in the env would make `type_of/1,2` much more powerful, but I guess we have to start somewhere. For now I added a comment that defining varbinds is not yet possible with `env/1,2`.